src/StringTools.UnitTests/OpportunisticIntern_Tests.cs

```diff
@@ -0,0 +1,142 @@
+        {
+            IsInternable(ref internable).ShouldBeFalse();
+        }
+
+        private static void AssertNotInternable(char[] ch)
+        {
+            AssertNotInternable(new InternableString(ch.AsSpan(0, ch.Length)));
+        }
+
+        protected static void AssertNotInternable(string value)
+        {
+            AssertNotInternable(value.ToCharArray());
+        }
+
+        /// <summary>
+        /// Test interning segment of char array
+        /// </summary>
+        [Fact]
+        public void SubArray()
+        {
+            var result = AssertInternable(new char[] { 'a', 't', 'r', 'u', 'e' }, 1, 4);
+            result.ShouldBe("true");
+        }
+
+        /// <summary>
+        /// Test interning segment of char array
+        /// </summary>
+        [Fact]
+        public void SubArray2()
+        {
+            var result = AssertInternable(new char[] { 'a', 't', 'r', 'u', 'e', 'x' }, 1, 4);
+            result.ShouldBe("true");
+        }
+
+        /// <summary>
+        /// This is the list of hard-coded interns. They should report interned even though they are too small for normal interning.
+        /// </summary>
+        [Fact]
+        public void KnownInternableTinyStrings()
+        {
+            AssertInternable("C#");
+            AssertInternable("F#");
+            AssertInternable("VB");
+            AssertInternable("True");
+            AssertInternable("TRUE");
+            AssertInternable("Copy");
+            AssertInternable("v4.0");
+            AssertInternable("true");
+            AssertInternable("FALSE");
+            AssertInternable("false");
+            AssertInternable("Debug");
+            AssertInternable("Build");
+            AssertInternable("''!=''");
+            AssertInternable("AnyCPU");
+            AssertInternable("Library");
+            AssertInternable("MSBuild");
+            AssertInternable("Release");
+            AssertInternable("ResolveAssemblyReference");
+        }
+
+        /// <summary>
+        /// Test a set of strings that are similar to each other
+        /// </summary>
+        [Fact]
+        public void InternableDifferingOnlyByNthCharacter()
+        {
+            string test = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890!@#$%^&*()_+ABCDEFGHIJKLMNOPQRSTUVabcdefghijklmnopqrstuvwxyz0150";
+            for (int i = 0; i < test.Length; ++i)
+            {
+                string mutated = test.Substring(0, i) + " " + test.Substring(i + 1);
+                AssertInternable(mutated);
+            }
+        }
+
+        /// <summary>
+        /// Test The empty string
+        /// </summary>
+        [Fact]
+        public void StringDotEmpty()
+        {
+            AssertInternable(string.Empty);
+        }
+
+        /// <summary>
+        /// Test an empty string.
+        /// </summary>
+        [Fact]
+        public void DoubleDoubleQuotes()
+        {
```

Comment:
Maybe a test that two strings that were initialized at different points come out to the same reference when interned if they happen to be identical with one (different) string between them that comes out to a different reference?

Like:
```
string foo = "bar";
string quux = "qux";
string baz = "bar";
ReferenceNotEquals(foo, baz); // Have to make them longer for this to be true?
foo = AssertInternable(foo); // Pretending it returns the string
quux = AssertInternable(quux);
baz = AssertInternable(baz);
ReferenceEquals(foo, baz);
ReferenceNotEquals(baz, quux);
```
?

Reply:
`foo` and `baz` in your code snipped are going to be the exact same instance thanks to runtime string interning so the first `ReferenceNotEquals` does not hold, regardless of how long the string is. I believe that a test like you're suggesting - passing two different references to the same contents and verifying that interning them results in the same reference - already exists in `WeakStringCache_Tests`.